### PR TITLE
bluetooth: GATT DM discover many service instances by UUID

### PR DESCRIPTION
**Pull request by KAGA164**
Adds a possibility to discover all the same service instances using the UUID

### CODE REVIEW

File: doc/nrf/releases/release-notes-changelog.rst [hunk 1]
```diff
@@ -153,6 +153,7 @@ Bluetooth libraries and services
 * :ref:`gatt_dm_readme` library:
 
   * Fixed discovery of empty services.
+  * Added possibility to discovering many the same services instances using the UUID.
```
Comment (peknis) [inline doc/nrf/releases/release-notes-changelog.rst:156]:
```suggestion
  * Added option to discover several service instances using the UUID.
```

File: include/bluetooth/gatt_dm.h [hunk 2]
```diff
@@ -292,6 +292,11 @@ int bt_gatt_dm_start(struct bt_conn *conn,
  * This function continues service discovery.
  * Call it after the previous data was released by @ref bt_gatt_dm_data_release.
  *
+ * When the service UUID had been set when service discovery was started using
+ * @ref bt_gatt_dm_start. This function discover the next instance of the same service
+ * if there are more instances on the GATT server. You can use this function to discover
+ * all services instances with given UUID.
```
Comment (peknis) [inline include/bluetooth/gatt_dm.h:298]: Maybe something like this:
The service UIDD is set when you start service discovery using the ... function. If there are more instances of the same service on the GATT server, it discovers the next instance. Use this function to discover all service instances with the given UIDD.

**Review by peknis — CHANGES_REQUESTED** [review]

File: doc/nrf/releases/release-notes-changelog.rst (hunk 1 above)
Comment (KAGA164) [reply doc/nrf/releases/release-notes-changelog.rst]: Done

File: include/bluetooth/gatt_dm.h (hunk 2 above)
Comment (KAGA164) [reply include/bluetooth/gatt_dm.h]: Ok, I changed it

File: include/bluetooth/gatt_dm.h [hunk 3]
```diff
@@ -292,6 +292,11 @@ int bt_gatt_dm_start(struct bt_conn *conn,
  * This function continues service discovery.
  * Call it after the previous data was released by @ref bt_gatt_dm_data_release.
  *
+ * If the service UUID is set when you start service discovery using the
```
Comment (peknis) [inline include/bluetooth/gatt_dm.h:295]: Please check again my suggestion. This does not really seem right yet. At least the first sentence is a bit weird.

**Review by peknis — CHANGES_REQUESTED** [review]

File: include/bluetooth/gatt_dm.h (hunk 3 above)
Comment (alwa-nordic) [reply include/bluetooth/gatt_dm.h:295]: I think we can remove this change to the doc. I don't think it adds any new information. If it's not clear that the service discovery will find all matching services, then we should clarify that in the doc of `bt_gatt_dm_start`.

**Review by alwa-nordic — DISMISSED** [review]

File: subsys/bluetooth/gatt_dm.c [hunk 4]
```diff
@@ -618,8 +629,16 @@ int bt_gatt_dm_start(struct bt_conn *conn,
 	dm->cur_attr_id = 0;
 	sys_slist_init(&dm->chunk_list);
 	dm->cur_chunk_len = 0;
+	dm->search_svc_by_uuid = svc_uuid ? 1UL : 0UL;
```
[the remark this answers is not in the document]
Comment (KAGA164) [reply subsys/bluetooth/gatt_dm.c]: Sure. Done

File: include/bluetooth/gatt_dm.h [hunk 5]
```diff
@@ -268,17 +268,26 @@ const struct bt_gatt_dm_attr *bt_gatt_dm_desc_next(
  * another one, wait for the result of the previous procedure to finish
  * and call @ref bt_gatt_dm_data_release if it was successful.
  *
+ * If the service UUID is set when you start service discovery using the
+ * @ref bt_gatt_dm_start function. If there are more instances of the same service
+ * on the GATT server, it discovers the next instance. Use this function to discover
+ * all service instances with the given UUID.
```
Comment (peknis) [inline include/bluetooth/gatt_dm.h:274]: This still does not make sense. The paragraph starts with "If...", but there is no explanation of what follows then.

**Review by peknis — CHANGES_REQUESTED** [review]

File: include/bluetooth/gatt_dm.h (hunk 5 above)
Comment (KAGA164) [reply include/bluetooth/gatt_dm.h:274]: Sorry I supposed that I removed this whole paragraph

Comment (KAGA164) [reply include/bluetooth/gatt_dm.h]: Done

File: include/bluetooth/gatt_dm.h [hunk 6]
```diff
@@ -272,13 +272,17 @@ const struct bt_gatt_dm_attr *bt_gatt_dm_desc_next(
...
+ * @param[in,out] context Context argument to be passed to
+ *                callback functions.
  *
  * @note
  * If @p svc_uuid is set to NULL, all services may be discovered.
  * To process the next service, call @ref bt_gatt_dm_continue.
  *
+ * If @p svc_uuid is set to the service UUID, all
+ * service instances may be discovered.
+ * To discover next service instance if it exist, call @ref bt_gatt_dm_continue.
```
Comment (peknis) [inline include/bluetooth/gatt_dm.h:284]:
```suggestion
 * Call @ref bt_gatt_dm_continue to discover the next service instance.
```

**Review by peknis — CHANGES_REQUESTED** [review]

File: include/bluetooth/gatt_dm.h (hunk 6 above)
Comment (KAGA164) [reply include/bluetooth/gatt_dm.h]: Done

**Review by dchat-nordic — APPROVED** [review]

**Review by doki-nordic — APPROVED** [review]

File: include/bluetooth/gatt_dm.h [hunk 7]
```diff
@@ -268,17 +268,26 @@ const struct bt_gatt_dm_attr *bt_gatt_dm_desc_next(
...
```
[the remark this answers is not in the document]
Comment (KAGA164) [reply include/bluetooth/gatt_dm.h]: This is already re-phrased

File: include/bluetooth/gatt_dm.h [hunk 8]
```diff
@@ -268,17 +268,26 @@ const struct bt_gatt_dm_attr *bt_gatt_dm_desc_next(
  * another one, wait for the result of the previous procedure to finish
  * and call @ref bt_gatt_dm_data_release if it was successful.
  *
+ * If the service UUID is set when you start service discovery using the
```
[the remark this answers is not in the document]
Comment (KAGA164) [reply include/bluetooth/gatt_dm.h]: This paragraph is removed now

**Review by grochu — APPROVED** [review]

**Comment by thst-nordic** [comment]
New test-spec is working correctly, so I will remove the redundant github labels.

**Review by grochu — APPROVED** [review]

**Pull request merged**